### ISSUE TITLE
Critical hashcode bug in pre sdk: generic types that are equal but not identical don't have the same hashcode

### ISSUE THREAD

**Issue by geovaly** [dart-lang/sdk#45060]
Hello,

### **Bug conditions:**

Flutter last version: Flutter 1.26.0-17.6.pre  Beta Channel with null safety
environment sdk: '>=2.12.0-29.10.beta <3.0.0', the bug does not occur in sdk: ">=2.7.0 <3.0.0"

### **Expected behavior:** 

2 objects that are equal ( == ) but not identical should have the same hashcode. This should be applied also for types because they are objects. For ex. my application is crashing when accessing a dictionary in the service locator (getIt) because of this bug.

### **Simplified code to show the bug:** 

```
import 'dart:developer';

void main() {
  var executor = new Executor();
  new Command1().execute(executor); //test failed
  executor.execute<Command1>(new Command1()); //test passed
}

class Command1 extends CommandBase<Command1> {
}

abstract class Command {
}

abstract class CommandBase<TSelf extends CommandBase<TSelf>> extends Command  {
  void execute(Executor e) {
    TSelf self = this as TSelf;
    e.execute<TSelf>(self);
  }
}

abstract class CommandHandler<T extends Command> {
  void handle(T command);
}

class Executor {
  void execute<T extends Command>(T action){
    testTypeEquality<CommandHandler<T>, CommandHandler<Command1>>();
  }
}

void testTypeEquality<T1,T2>(){
  var hash1 = T1.hashCode;
  var hash2 = T2.hashCode;
  var equals = T1 == T2;
  var isIdentical = identical(T1, T2);
  log("hash1: $hash1");
  log("hash2: $hash2");
  log("T1 equals T2: $equals");
  log("T1 identical T2: $isIdentical");
  log("*******");
}
```

[main.txt](https://github.com/dart-lang/sdk/files/6013821/main.txt)

### Output:

**[log] hash1: 901285914
[log] hash2: 94173417
[log] T1 equals T2: true
[log] T1 identical T2: false**
[log] *******
[log] hash1: 94173417
[log] hash2: 94173417
[log] T1 equals T2: true
[log] T1 identical T2: true
[log] *******

**Congratulation** for flutter and dart.. they are amazing. Looking forward for null safety release.

**Comment by crelier** [comment]
CL is under review: https://dart-review.googlesource.com/c/sdk/+/186421

**Comment by geovaly** [comment]
Hi. Do you have an idea when the fix it will be in the sdk release or prerelease?

**Comment by crelier** [comment]
The fix missed the branch date for stable Dart 2.12 (Flutter 2.0) by a few days and was not judged critical enough to be cherry picked, sorry. But it will definitely be included in Dart 2.13 (Flutter 2.1), either dev, beta, or stable.

**Comment by geovaly** [comment]
Thank you Regis for the update.
